integrations/spotter.c

```diff
@@ -0,0 +1,206 @@
+ @return BmErr if unsuccessful
+*/
+BmErr bm_fprintf(uint64_t target_node_id, const char *file_name,
+                 uint8_t print_time, const char *format, ...) {
+  BmErr err = BmOK;
+  bm_print_publication_t *printf_pub = NULL;
+  va_list va;
+
+  do {
+    va_start(va, format);
+    // check how long the string we are printing will be
+    int32_t data_len = vsnprintf(NULL, 0, format, va);
+    va_start(va, format);
+    if (data_len == 0) {
+      err = BmENODATA;
+      break;
+    }
+
+    int32_t fname_len = 0;
+    if (file_name) {
+      fname_len = bm_strnlen(file_name, max_file_name_len);
+      if (fname_len >= max_file_name_len) {
+        err = BmEMSGSIZE;
+        break;
+      }
+    }
+
+    if (data_len > max_str_len(fname_len)) {
+      err = BmEMSGSIZE;
+      break;
+    }
+
+    err = BmENOMEM;
+    size_t printf_pub_len =
+        sizeof(bm_print_publication_t) + data_len + fname_len;
+    printf_pub = (bm_print_publication_t *)bm_malloc(printf_pub_len);
+
+    if (printf_pub) {
+      err = BmOK;
+      memset(printf_pub, 0, printf_pub_len);
+      printf_pub->target_node_id = target_node_id;
+      printf_pub->fname_len = fname_len;
+      printf_pub->data_len = data_len;
+      printf_pub->print_time = print_time;
+
+      if (file_name) {
+        memcpy(printf_pub->fnameAndData, file_name, fname_len);
+      }
+
+      data_len +=
+          1; // add one so vsnprintf doesn't overwrite the last character with null
+      // do this after we malloc a buffer, so the extra null after the last character doesn't print to the file/ascii terminal
+      printf("Result: %d %d\n", fname_len, data_len - 1);
+      int32_t res = vsnprintf((char *)&printf_pub->fnameAndData[fname_len],
+                              data_len, format, va);
+      printf("Result: %d %d %s\n", res, data_len - 1, (char *)&printf_pub->fnameAndData[fname_len]);
+      if (res < 0 || (res != data_len - 1)) {
+        err = BmEBADMSG;
+        break;
+      }
+
+      if (file_name) {
+        if (!bm_pub(SPOTTER_FPRINTF_TOPIC, printf_pub, printf_pub_len,
+                    FPRINTF_TYPE, BM_COMMON_PUB_SUB_VERSION)) {
+          err = BmENETDOWN;
+        }
+      } else {
+        if (!bm_pub(SPOTTER_PRINTF_TOPIC, printf_pub, printf_pub_len,
+                    PRINTF_TYPE, BM_COMMON_PUB_SUB_VERSION)) {
+          err = BmENETDOWN;
+        }
+      }
+    }
+  } while (0);
+
+  va_end(va);
+
+  if (printf_pub) {
+    bm_free(printf_pub);
+    printf_pub = NULL;
+  }
+
+  return err;
+}
+
+/*!
+ @brief Bristlemouth generic append to file
+
+ @param[in] target_node_id - node_id to send to (0 = all nodes), the accept if it is
+                             subscribed to the topic that the printf is publishing to
+ @param[in] file_name - file name to print to
+ @param[in] *buf - buffer to send
+
+ @return BmOk if successful
+ @return BmErr if unsuccessful
+*/
+BmErr bm_file_append(uint64_t target_node_id, const char *file_name,
```

Comment:
What!? 🤦‍♂️ I did not know this existed. I see it's been there since June 2023. This would have been a better solution for https://github.com/bristlemouth/bm_protocol/issues/154. Oh well. 🤷‍♂️

Although interestingly, this is NOT specific to Spotter! The published topic is simply `fappend`.

I created backlog story 209609 for us to implement this on Spotter. It doesn't work right now because the implementation was started in an exploratory "how should Bristlemouth SD card writing work" kind of way, but it has not been finished.

This function should be removed from the Spotter integration file because (1) it doesn't currently work and (2) it's not Spotter-specific. It should be kept though, moved somewhere else. It doesn't seem to belong with the core middleware implementation. I'd propose maybe `file_ops.{c,h}` here in the integration folder? This is like was @victorsowa12 was proposing in our discussion on Monday. I wouldn't want to name it `file.h` or `files.h` because that seems too likely to conflict with other libraries. Thoughts?

Reply:
I really like that. I believe it is generic and would fit better elsewhere!